src/api/messages/getTemporaryFileUrl.js

```diff
@@ -0,0 +1,24 @@
+/* @flow strict-local */
+import type { Auth, ApiResponseSuccess } from '../transportTypes';
+import { apiGet } from '../apiFetch';
+
+type ApiResponseTempFileUrl = {|
+  ...ApiResponseSuccess,
+  url: string,
+|};
+
+/**
+ * This endpoint, introduced in zulip/zulip@9e72fbe3f, returns a partial URL
+ * string without the realm domain included in it. This URL allows temporarily
```

Comment:
Good to have these details in a jsdoc!

I'm struggling to find an example in our API bindings, but Greg guided me toward some conventions we use, when I was writing `src/utils/zuilpVersion.js` (you could take a look at that file):

1. Start with a terse single-line summary of what the thing is/does.
2. Then some summary, which gives relevant background, or places you might want to use the thing, or reasons to use it instead of something else, etc.

Good to include the details in `* @param filePath`, I think. Not sure what to do about `auth`; it's boring, since all the other bindings have it too. Maybe just an empty `@param auth`, with no explanation? Order matters, so I think there's an argument for putting something minimal before `@param filePath`.

Reply:
I'm not sure (and it doesn't seem to be documented in the zulip.yaml file you link to) whether the backend implementation ended up giving us a short-lived URL that was

1. just completely random, or
2. based on a combination of the user (even their private API key?) and the requested document, with, I guess, 
     2a. some randomness on top of that, or
     2b. completely determined by this info combined with a timestamp?

Possibly, the answer to this won't ever be important for the mobile app to be aware of, or these are considered implementation details by the server, so could change at any time. Assuming one of these is true, it seems appropriate for us to be intentionally vague about those details. The important piece is that the URL won't be valid after, I think, 60 seconds or so.

Reply:
I think there's even more detail about why we're using this endpoint, which would be good to include. See Greg's points [here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/view.20uploaded.20files/near/819779):
> The goal basically is to get
> * (a) a plain old URL that can be passed to a browser (not a URL plus some cookies or other headers that need to be set)
> * (b) that gets the user the file immediately, without requiring a fresh login
> * (c) but doesn't include a secret with much wider powers (i.e. their API key), which would risk a nasty surprise if they innocently show someone else the file by sending them the link.

I think (a) is explained by "without sending auth headers", but let's get (b) and (c) in there too.

Reply:
> I'm not sure (and it doesn't seem to be documented in the zulip.yaml file you link to) whether the backend implementation ended up giving us a short-lived URL that was
> 
> 1. just completely random, or
> 2. based on a combination of the user (even their private API key?) and the requested document, with, I guess,
>    2a. some randomness on top of that, or
>    2b. completely determined by this info combined with a timestamp?

I believe the implementation ended up using Django's `django.core.signing`, either its `dumps` function or the underlying `TimestampSigner`. There's docs here:
https://docs.djangoproject.com/en/3.0/topics/signing/
and the cryptographic concept it implements is an "HMAC":
https://en.wikipedia.org/wiki/HMAC

Basically we take the IDs of the user and the requested document, make a sentence saying "As of TIME, user USER can access file FILE", and then compute a signature. The URL contains that sentence and signature.

The signature is done with a secret key belonging to the server, and can be verified with the same secret key by just re-computing what the signature should be.

The HMAC is a common technique for webapps to hand the client some piece of data, especially this kind of access token, in what amounts to a transparent but tamper-proof envelope so the client can hand it back to the server later and the server can trust the contents. Just as it did for us here, this technique saves the server from having to store the tokens itself.


> Assuming one of these is true, it seems appropriate for us to be intentionally vague about those details.

Yeah, I think those are implementation details the client needn't care about, and the other points you mentioned are more relevant.

Reply:
Thanks!